Mathlib/InformationTheory/BinaryEntropy.lean

```diff
@@ -0,0 +1,345 @@
+    · by_cases pgthalf : 1/2 < p
+      · have := h2_lt_one_of_p_gt_half pgthalf ple1
+        linarith
+      · linarith
+  · simp [h]
+
+lemma h2_le_1 {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) : h₂ p ≤ 1 := by
+  by_cases hh: p = 1/2
+  · simp [*];
+  · by_cases h₂ p = 1
+    · simp [*]
+    · by_cases hhh: p < 1/2
+      · linarith [h2_lt_1_of_p_lt_half pge0 hhh]
+      · have : 1/2 < p := by
+          refine Ne.lt_of_le ?_ ?_
+          aesop
+          aesop
+        have := h2_lt_one_of_p_gt_half this ple1
+        exact LT.lt.le this
+
+---------------------------------------------------------------------------------- derivatives
+
+/-- Derivative of binary entropy function (shown in `deriv_h₂`) -/
+protected noncomputable def h₂deriv (p : ℝ) : ℝ := log₂ (1 - p) - log₂ p
+
+@[simp] lemma deriv_one_minus (x : ℝ) : deriv (fun (y : ℝ) ↦ 1 - y) x = -1 := by
+  have onem (y : ℝ) : 1 - y = -(y + -1) := by ring
+  simp_rw [onem]
+  simp only [neg_add_rev, neg_neg, differentiableAt_const, deriv_const_add', deriv_neg'']
+
+@[simp] lemma differentiable_1_minusp (p : ℝ) : DifferentiableAt ℝ (fun p => 1 - p) p := by
+  have (p : ℝ) : 1 - p = -(p - 1) := by ring
+  simp_rw [this]
+  apply differentiableAt_neg_iff.mpr
+  apply DifferentiableAt.add_const
+  simp only [differentiableAt_id']
+
+-- TODO don't need assumptions
+lemma deriv_log_one_sub {x : ℝ} (hh : x ≠ 1): deriv (fun p ↦ log (1 - p)) x = -(1-x)⁻¹ := by
+  rw [deriv.log]
+  simp only [deriv_one_minus]
+  field_simp
+  exact differentiable_1_minusp x
+  exact sub_ne_zero.mpr hh.symm
+
+@[simp] lemma differentiableAt_log_const_neg {x c : ℝ} (h : x ≠ c) :
+    DifferentiableAt ℝ (fun p ↦ log (c - p)) x := by
+  apply DifferentiableAt.log
+  apply DifferentiableAt.sub
+  apply differentiableAt_const
+  apply differentiableAt_id'
+  exact sub_ne_zero.mpr (id (Ne.symm h))
+
+-- TODO don't need assumptions
+lemma deriv_h₂' {x : ℝ} (h: x ≠ 0) (hh : x ≠ 1) :
+    deriv (fun p => -p * log₂ p - (1 - p) * log₂ (1 - p)) x = log₂ (1 - x) - log₂ x := by
+  simp [log₂, logb]
+  rw [deriv_sub, deriv.neg]
+  simp_rw [mul_div]
+  rw [deriv_div_const, deriv_mul_log h, deriv_div_const]
+  simp_rw [mul_sub_right_distrib]
+  simp only [one_mul]
+  rw [deriv_sub, deriv_log_one_sub hh]
+  · rw [deriv_mul, deriv_id'']
+    rw [deriv.log]
+    simp only [one_mul, deriv_one_minus]
+    field_simp
+    ring_nf
+    calc -1 + (-log x - x * (1 - x)⁻¹) + (1 - x)⁻¹ + log (1 - x)
+      _ = -1 + (1 - x) * (1 - x)⁻¹ + log (1 - x) - log x := by ring
+      _ = -log x + log (1 - x) := by
+        field_simp [sub_ne_zero.mpr hh.symm]
+        ring
+    apply differentiable_1_minusp
+    exact sub_ne_zero.mpr hh.symm
+    apply differentiableAt_id'
+    exact differentiableAt_log_const_neg hh
+  · exact differentiableAt_log_const_neg hh
+  · apply DifferentiableAt.mul
+    apply differentiableAt_id'
+    apply DifferentiableAt.log
+    exact differentiable_1_minusp x
+    exact sub_ne_zero.mpr hh.symm
+  · apply DifferentiableAt.neg
```

Comment:
Does `fun_prop` help in here?

Reply:
I don't know how to use `fun_prop`. Seems like a very recent addition (It's not even in ["the tactics list"](https://github.com/haruhisa-enomoto/mathlib4-all-tactics)).

I tried to stick it into various places (had to  rebase on origin/main to access it) but to no avail so far.

Reply:
I think it should be doable, just `import Mathlib.Tactic.FunProp.Differentiable` and the try the proof with `fun_prop (disch:=aesop)`.

Reply:
Now it works in some cases where I can replace *one* individual `apply X` or `exact X` by `fun_prop (disch:=aesop)`. I haven't been able to use it to replace more than one line. Is it still worth using? It might make the proof brittle when `fun_prop` or `aesop` are updated in the future. Basically, this problem is analogous to "non-terminal simp"?

My file has a lot of divisions by `log 2` (binary logarithms) which is bad for the automated tooling (because it needs to know `log 2 ≠ 0`). Maybe I should have done base-e logs instead. On the other hand, I prefer measuring entropy in bits and don't want to make this decision just based on pleasing the automation.

Reply:
It should be just one call to `fun_prop`. I will give it a try if you update the branch to a mathlib version that has `fun_prop`.

Reply:
@lecopivo  Updated the branch.

Reply:
The failed CI is just "unused import" because I deleted my attempts at using `fun_prop`...

Reply:
A useful strategy is to run `fun_prop (disch:=trace_state;sorry)`. 

Then you look at the info view see what was sorried to close the goal. Now you can add these statement to the local context and use `fun_prop (disch:=assumption)`

Reply:
This is interesting. Feel free to push to the branch. I will not do any further work on this PR until b-mehta compares with their code.

Reply:
I will do that once I port `fun_trans` tactic and the proof will be just few lines.